src/psykal/tra_adv_alg.F90

```diff
@@ -141,36 +145,87 @@ PROGRAM tra_adv
 
   call timer_stop(step_timer)
 
-  OPEN(unit = 4, file = 'output.dat', form='formatted')
-  
-  DO jk = 1, jpk-1
-     DO jj = 2, jpj-1
-        DO ji = 2, jpi-1
-           write(4,*) mydomain(ji,jj,jk)
-        END DO
-     END DO
-  END DO
-
-  CLOSE(4)
-
-  DEALLOCATE( mydomain )
-  DEALLOCATE( zwx )
-  DEALLOCATE( zwy )
-  DEALLOCATE( zslpx )
-  DEALLOCATE( zslpy )
-  DEALLOCATE( pun )
-  DEALLOCATE( pvn )
-  DEALLOCATE( pwn )
-  DEALLOCATE( umask)
-  DEALLOCATE( vmask)
-  DEALLOCATE( tmask)
-  DEALLOCATE( zind )
-  DEALLOCATE( ztfreez )
-  DEALLOCATE( rnfmsk)
-  DEALLOCATE( upsmsk)
-  DEALLOCATE( rnfmsk_z)
-  DEALLOCATE( tsn)
+  call write_field(mydomain)
 
   CALL timer_report()
 
+CONTAINS
+
+  subroutine init_fields()
+    integer  :: jpi, jpj, jpk
+    REAL(wp) :: r
+    integer  :: ji, jj, jk
+
+    jpi = mydomain%grid%simulation_domain%xstop
+    jpj = mydomain%grid%simulation_domain%xstop
+    jpk = mydomain%grid%nlevels
+
+    ! Array initialization
+    r = jpi*jpj*jpk
+
+    ! the following three lines can be uncommented to randomize the array
+    ! initialization
+    !call random_seed()
+    !call random_number(r)
+    !r = r*jpi*jpj*jpk
+
+   DO jk = 1, jpk
+      DO jj = 1, jpj
+          DO ji = 1, jpi
+              mydomain%data(ji,jj,jk) =ji*jj*jk/r
+              pun%data(ji,jj,jk) =ji*jj*jk/r
+              pvn%data(ji,jj,jk) =ji*jj*jk/r
+              pwn%data(ji,jj,jk) =ji*jj*jk/r
+              tsn%data(ji,jj,jk)= ji*jj*jk/r
+              ! Set the mask values to match those in the original
+              ! benchmark harness. The t-mask values would normally
+              ! be read from file and then the values of the other
+              ! masks computed from them.
+              tsn%grid%tmask(ji,jj,jk) = ji*jj*jk/r
+              tsn%grid%umask(ji,jj,jk) = ji*jj*jk/r
+              tsn%grid%vmask(ji,jj,jk) = ji*jj*jk/r
+          END DO
+      END DO
+   END DO
+
+   r = jpi*jpj
+   DO jj=1, jpj
+      DO ji=1, jpi
+         ztfreez%data(ji,jj) = ji*jj/r
+         upsmsk%data(ji,jj) = ji*jj/r
+         rnfmsk%data(ji,jj) = ji*jj/r
+      END DO
+   END DO
+
+   DO jk=1, jpk
+      rnfmsk_z%data(jk)=jk/jpk
+   END DO
+
+  end subroutine init_fields
+
+  subroutine write_field(mydomain)
```

Comment:
Agreed. The infrastructure does already have some limited support for I/O but I opted for minimal changes here (because I wanted to be able to just diff the outputs of the psykal and original codes...)